### ISSUE TITLE
"this.r is not a function"

### ISSUE THREAD

**Issue by onlywei** [standard-things/esm#317]
Hi, I am trying to interop use this with an older version of @std/esm in a package that I wrote. Here's the setup:

In my application, I have `esm@3.0.5` and I'm starting the application using the command:

```node -r esm index.js```

Inside `index.js`, we have this:

```js
import something from 'my-package';
```

**When I try to start my application, I get the error**:

```
file:///path/to/myapp/node_modules/path-to-pacakge/index.mjs:1
TypeError: this.r is not a function
    at Object.<anonymous> (file:///path/to/myapp/node_modules/path-to-pacakge/index.mjs:1)
    at E (file:///path/to/myapp/node_modules/registry-path/@std/esm/0.25.2/node_modules/@std/esm/esm.js:1)
    at Object.run (file:///path/to/myapp/node_modules/registry-path/@std/esm/0.25.2/node_modules/@std/esm/esm.js:1)
    at Object.<anonymous> (file:///path/to/myapp/node_modules/path-to-pacakge/index.mjs:1)
    at Module.e (file:///path/to/myapp/node_modules/registry-path/@std/esm/0.25.2/node_modules/@std/esm/esm.js:1)
    at X (file:///path/to/myapp/node_modules/registry-path/@std/esm/0.25.2/node_modules/@std/esm/esm.js:1)
    at /path/to/myapp/node_modules/registry-path/@std/esm/0.25.2/node_modules/@std/esm/esm.js:1
    at module.exports.r.a (file:///path/to/myapp/node_modules/registry-path/@std/esm/0.25.2/node_modules/@std/esm/esm.js:1)
    at Module._ (file:///path/to/myapp/node_modules/registry-path/@std/esm/0.25.2/node_modules/@std/esm/esm.js:1)
    at Module.d._compile (file:///path/to/myapp/node_modules/registry-path/@std/esm/0.25.2/node_modules/@std/esm/esm.js:1)
```

"my-package" is a package that is published to our private npm enterprise. It has `@std/esm@0.25.2` installed. Its `package.json` file looks like this:
```json
{
  "main": "index.js",
  "module": "index.mjs",
  "esnext": "index.mjs",
  "@std/esm": "cjs"
}
```

The `index.js` file looks like this:

```js
module.exports = require('@std/esm')(module)('./index.mjs').default;
```

Am I doing this inter-op wrong?

**Comment by jdalton** [comment]
Hi @onlywei!

Do you have an example repro repo I can look at?

**Comment by onlywei** [comment]
@jdalton unfortunately it's a little difficult for me to create an example repro repo since I'm using a private NPM enterprise that's not accessible to the public for my package.

Is there something that seems off to you with the details I've provided?

**Comment by jdalton** [comment]
No worries. I was able to reproduce on my end with another package. I'm digging into it. It looks like the issue is as follows:

When `esm` loads files it is-the-load-pipeline. So when `@std/esm` sets itself up it will use the existing `esm` pipe-line thinking it's Node's built-in. This means that `esm` is being used to process `@std/esm` (not expected) and this causes the issue. I'm going to dig in a bit more and see where they disconnect is and also come up with a solution to avoid the issue in the first place.

**Comment by jdalton** [comment]
Ok, I have a patch though I'm not super happy with it at the moment. I'm going to simmer on it for a bit before pushing it.

**Comment by jdalton** [comment]
The gist is this.
Using the isolated per package `require = require("@std/esm")(module)` is great!
It isolates the loader from clobbering others. But when used with the `node -r` flag it attaches to the global hook since this allows scripts not actively opting-in to get ESM support. The global hook is susceptible to collisions and in this case, of different packages and different versions, it hit.

**Comment by jdalton** [comment]
**Update:**

I'll patch release both `@std/esm` and `esm` to resolve this.

**Comment by jdalton** [comment]
Patched https://github.com/standard-things/esm/commit/158f61d1192cdc0f781b666b57b2f962e52e2a38 and https://github.com/standard-things/esm/commit/73c707473fa03b03a09e6dc067e29bf55462093c. I'll publish updates in the morning.

**Comment by jdalton** [comment]
`@std/esm` has been updated to [v0.25.3](https://github.com/standard-things/esm/releases/tag/0.25.3) and `esm` has been updated to [v3.0.6](https://github.com/standard-things/esm/releases/tag/3.0.6) 🎉.

**Comment by onlywei** [comment]
Great! Thanks @jdalton

**Comment by onlywei** [comment]
@CacheControl I haven't been able to test it yet, but I imagine when I do that I would have to update both my app and package.

**Comment by jdalton** [comment]
@CacheControl I patched the issue for `@std/esm` v0.25.3 so you'll still run into the issue with older `@std/esm` versions for sure. I also only smoothed over the case for when using `esm` to load packages that currently use `@std/esm` and not for `@std/esm` loading packages that use `esm` because `@std/esm` is pre-stable and should be migrated from asap.

**Comment by CacheControl** [comment]
@jdalton thanks for the clarification.  I shall migrate everything to `esm` asap!
